### PR TITLE
added totaltext recog converter

### PR DESCRIPTION
**Pull request by quincylin1**
added totaltext converter to generate cropped images and ann .txt file

### CODE REVIEW

File: tools/data/textrecog/totaltext_converter.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        files(list): The list of tuples (img_file, groundtruth_file)
+    """
+    assert isinstance(img_dir, str)
+    assert img_dir
+    assert isinstance(gt_dir, str)
+    assert gt_dir
+
+    # note that we handle png and jpg only. Pls convert others such as gif to
+    # jpg or png offline
+    suffixes = ['.png', '.PNG', '.jpg', '.JPG', '.jpeg', '.JPEG']
```
Comment (innerlee) [inline tools/data/textrecog/totaltext_converter.py:36]: you can compare the lowercased suffixes

File: tools/data/textrecog/totaltext_converter.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+        imgs_list.extend(glob.glob(osp.join(img_dir, '*' + suffix)))
+
+    imgs_list = [
+        drop_orientation(f) if is_not_png(f) else f for f in imgs_list
+    ]
+
+    files = []
+    if split == 'training':
+        for img_file in imgs_list:
+            gt_file = gt_dir + '/poly_gt_' + osp.splitext(
```
Comment (cuhk-hbsun) [inline tools/data/textrecog/totaltext_converter.py:50]:
```suggestion
            gt_file = osp.join(gt_dir, 'poly_gt_' + osp.splitext(
```

File: tools/data/textrecog/totaltext_converter.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+                osp.basename(img_file))[0] + '.mat'
```
Comment (cuhk-hbsun) [inline tools/data/textrecog/totaltext_converter.py:51]:
```suggestion
                osp.basename(img_file))[0] + '.mat')
```

File: tools/data/textrecog/totaltext_converter.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+            # gt_file = gt_dir + '/' + osp.splitext(
```
Comment (cuhk-hbsun) [inline tools/data/textrecog/totaltext_converter.py:52]: del this line

File: tools/data/textrecog/totaltext_converter.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+            #    osp.basename(img_file))[0] + '.png'
```
Comment (cuhk-hbsun) [inline tools/data/textrecog/totaltext_converter.py:53]: del this line

File: tools/data/textrecog/totaltext_converter.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+            files.append((img_file, gt_file))
+        assert len(files), f'No images found in {img_dir}'
+        print(f'Loaded {len(files)} images from {img_dir}')
+    elif split == 'test':
+        for img_file in imgs_list:
+            gt_file = gt_dir + '/poly_gt_' + osp.splitext(
```
Comment (cuhk-hbsun) [inline tools/data/textrecog/totaltext_converter.py:59]:
```suggestion
            gt_file = osp.join(gt_dir, 'poly_gt_' + osp.splitext(
```

File: tools/data/textrecog/totaltext_converter.py [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+                osp.basename(img_file))[0] + '.mat'
```
Comment (cuhk-hbsun) [inline tools/data/textrecog/totaltext_converter.py:60]:
```suggestion
                osp.basename(img_file))[0] + '.mat')
```

File: tools/data/textrecog/totaltext_converter.py [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+
+    contours = []
+    words = []
+    data = scio.loadmat(gt_path)
+    if split == 'training':
+        data_polygt = data['polygt']
+    elif split == 'test':
+        data_polygt = data['polygt']
+
+    for i, lines in enumerate(data_polygt):
```
Comment (cuhk-hbsun) [inline tools/data/textrecog/totaltext_converter.py:117]:
```suggestion
    for lines in data_polygt:
```

File: tools/data/textrecog/totaltext_converter.py [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+        anno = dict(word=word, bbox=bbox)
+        anno_info.append(anno)
+
+    img_info.update(anno_info=anno_info)
+    return img_info
+
+
+def generate_ann(root_path, split, image_infos):
+
+    dst_image_root = osp.join(root_path, 'dst_imgs/' + split)
```
Comment (cuhk-hbsun) [inline tools/data/textrecog/totaltext_converter.py:180]:
```suggestion
    dst_image_root = osp.join(root_path, 'dst_imgs', split)
```

File: tools/data/textrecog/totaltext_converter.py [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+    dst_label_file = osp.join(root_path, 'train_label.txt')
+    os.makedirs(dst_image_root, exist_ok=True)
+
+    lines = []
+    for image_info in image_infos:
+        index = 1
+        src_img_path = osp.join(root_path + '/imgs', image_info['file_name'])
```
Comment (cuhk-hbsun) [inline tools/data/textrecog/totaltext_converter.py:187]:
```suggestion
        src_img_path = osp.join(root_path, 'imgs', image_info['file_name'])
```

File: tools/data/textrecog/totaltext_converter.py [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+        image = mmcv.imread(src_img_path)
+        src_img_root = osp.splitext(image_info['file_name'])[0].split('/')[1]
+
+        for anno in image_info['anno_info']:
+            word = anno['word']
+            dst_img = crop_img(image, anno['bbox'])
+            dst_img_name = f'{src_img_root}_{index}' + '.png'
```
Comment (cuhk-hbsun) [inline tools/data/textrecog/totaltext_converter.py:194]:
```suggestion
            dst_img_name = f'{src_img_root}_{index}.png'
```

File: tools/data/textrecog/totaltext_converter.py [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+            index += 1
+            dst_img_path = osp.join(dst_image_root, dst_img_name)
+            cv2.imwrite(dst_img_path, dst_img)
```
Comment (cuhk-hbsun) [inline tools/data/textrecog/totaltext_converter.py:197]:
```suggestion
            mmcv.imwrite(dst_img_path, dst_img)
```

File: tools/data/textrecog/totaltext_converter.py [hunk 13]
```diff
@@ -0,0 +1,284 @@
+import argparse
+import glob
+import os
+import os.path as osp
+from functools import partial
+
+import cv2
```
Comment (cuhk-hbsun) [inline tools/data/textrecog/totaltext_converter.py:7]: del this line

**Review by cuhk-hbsun — APPROVED** [review]

**Pull request merged**